File: ddsm115/run_motor.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from std_srvs.srv import Empty
from .ddsm115_operator_def import MOTOR_COMMAND
import time

class MotorControlNode(Node):
    def __init__(self):
        super().__init__('motor_control')
        self.MOTOR1 = MOTOR_COMMAND(0)
        
        # "/direction" 토픽 구독 설정
        self.subscription = self.create_subscription(
            String,
            '/direction',
            self.direction_callback,
            10)
        
    def direction_callback(self, msg):
        direction = msg.data
        self.get_logger().info(f'Received direction: {direction}')
        
        # 방향에 따른 모터 속도 조절
        if direction == 'Turn Left':
            # 모터 속도 조절 로직 구현
            self.MOTOR1.SET_VELOCITY(ID=1, SPEED=50)
            time.sleep(0.005)
            self.MOTOR1.SET_VELOCITY(ID=2, SPEED=100)
            time.sleep(0.005)
        elif direction == 'Go Straight':
            # 모터 속도 조절 로직 구현
            self.MOTOR1.SET_VELOCITY(ID=1, SPEED=100)
            time.sleep(0.005)
            self.MOTOR1.SET_VELOCITY(ID=2, SPEED=100)
            time.sleep(0.005)
        elif direction == 'Turn Right':
            # 모터 속도 조절 로직 구현
            self.MOTOR1.SET_VELOCITY(ID=1, SPEED=100)
            time.sleep(0.005)
            self.MOTOR1.SET_VELOCITY(ID=2, SPEED=50)
            time.sleep(0.005)
        elif direction == 'Go Back':
            # 모터 속도 조절 로직 구현
            self.MOTOR1.SET_VELOCITY(ID=1, SPEED=-100)
            time.sleep(0.005)
            self.MOTOR1.SET_VELOCITY(ID=2, SPEED=-100)
            time.sleep(0.005)
        elif direction == 'Stop':
            # 모터 속도 조절 로직 구현
            self.MOTOR1.SET_VELOCITY(ID=1, SPEED=0)
            time.sleep(0.005)
            self.MOTOR1.SET_VELOCITY(ID=2, SPEED=0)
            time.sleep(0.005)
        else:
            self.get_logger().info('Unknown direction received.')

def main(args=None):
    rclpy.init(args=args)
    motor_control_node = MotorControlNode()
    try:
        rclpy.spin(motor_control_node)
    except KeyboardInterrupt:
        pass
    finally:
        motor_control_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: run_motor.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "ddsm_cpp_py/motor_command.hpp" // 올바른 경로로 수정
#include <chrono>
#include <memory>
#include <thread>

using namespace std::chrono_literals;

class MotorControlNode : public rclcpp::Node {
public:
    MotorControlNode() : Node("motor_control"), motor1(0), motor2(1), motor3(2), motor4(3) {
        subscription = this->create_subscription<std_msgs::msg::String>(
            "/direction",
            10,
            std::bind(&MotorControlNode::direction_callback, this, std::placeholders::_1));
    }

private:
    void direction_callback(const std_msgs::msg::String::SharedPtr msg) {
        auto direction = msg->data.c_str();
        RCLCPP_INFO(this->get_logger(), "Received direction: %s", direction);

        if (std::string(direction) == "Turn Left") {
            // motor1.SET_VELOCITY(1, -50);
            // std::this_thread::sleep_for(5ms);
            // motor1.SET_VELOCITY(2, -50);
            // std::this_thread::sleep_for(5ms);
        } else if (std::string(direction) == "Go Straight") {
            motor1.SET_VELOCITY(1, -50);
            std::this_thread::sleep_for(5ms);
            motor1.SET_VELOCITY(2, 50);
            std::this_thread::sleep_for(5ms);
        } else if (std::string(direction) == "Turn Right") {
            // motor1.SET_VELOCITY(1, 50);
            // std::this_thread::sleep_for(5ms);
            // motor1.SET_VELOCITY(2, 0);
            // std::this_thread::sleep_for(5ms);
        } else if (std::string(direction) == "Go Back") {
            motor1.SET_VELOCITY(1, 50);
            std::this_thread::sleep_for(5ms);
            motor1.SET_VELOCITY(2, -50);
            std::this_thread::sleep_for(5ms);
        } else if (std::string(direction) == "Stop") {
            motor1.SET_VELOCITY(1, 0);
            std::this_thread::sleep_for(5ms);
            motor1.SET_VELOCITY(2, 0);
            std::this_thread::sleep_for(5ms);
        } else {
            RCLCPP_INFO(this->get_logger(), "Unknown direction received.");
        }
    }

    MOTOR_COMMAND motor1, motor2, motor3, motor4;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MotorControlNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
